### PR TITLE
add markdown render, if rep is text/markdown, render the md to html

### PR DESCRIPTION
**Pull request by YuqiXiao**
mod_markdown to process markdown response

### CODE REVIEW

**Comment by blackstorm** [comment]
is a plugin ?

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+		m.state.ReqRenderSuccess.Inc(1)
+		return bfe_module.BfeHandlerGoOn
+	}
+	return bfe_module.BfeHandlerGoOn
+}
+
+func (m *ModuleMarkdown) renderMarkDown(res *bfe_http.Response) error {
+	src, err := ioutil.ReadAll(res.Body)
+	if err != nil {
```
Comment (iyangsj) [inline bfe_modules/mod_markdown/mod_markdown.go:142]: 1. 需将返回的响应改为500响应？
2. 需统计错误指标

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		return err
+	}
+
+	dst, err := Render(src)
+	if err != nil {
```
Comment (iyangsj) [inline bfe_modules/mod_markdown/mod_markdown.go:147]: 同上

File: bfe_modules/mod_markdown/md_render.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+package mod_markdown
+
+import (
+	"fmt"
+)
+
+import (
+	"github.com/microcosm-cc/bluemonday"
+	"github.com/russross/blackfriday/v2"
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/md_render.go:23]: used in line 36

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	confPath := bfe_module.ModConfPath(cr, m.Name())
+	if m.conf, err = ConfLoad(confPath, cr); err != nil {
+		return fmt.Errorf("%s: conf load err %v", m.name, err)
+	}
+	openDebug = m.conf.Log.OpenDebug
+
+	if err = m.loadProductRuleConf(nil); err != nil {
+		return fmt.Errorf("%s: loadProductRuleConf() err %v", m.Name(), err)
+	}
+
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go:204]: done

File: bfe_modules/mod_markdown/md_rule_load.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func ProductRuleConfLoad(fileName string) (productRuleConf, error) {
+	var confFile productRuleConfFile
+	var conf productRuleConf
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		return conf, err
+	}
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/md_rule_load.go:142]: fix in commit  dd7abf9

File: bfe_modules/mod_markdown/md_rule_load.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+type productRules map[string]*markdownRules
+
+type productRuleConfFile struct {
+	Version string
+	Config  *productRulesFile
+}
+
+type productRuleConf struct {
+	Version string
+	Config  *productRules
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/md_rule_load.go]: fix in commit  dd7abf9

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	for _, rule := range *rules {
+		if !(*rule.Cond).Match(req) {
+			continue
+		}
+		m.state.ReqMarkDownRuleHit.Inc(1)
+
+		err := m.checkResponse(res)
+		if err != nil {
+			m.state.ReqRenderFailure.Inc(1)
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go]: fix in commit  dd7abf9

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+		err = m.renderMarkDown(res)
+		if err != nil {
+			if openDebug {
+				log.Logger.Debug("err in ModuleMarkdown.renderMarkDown(): %s", err)
+			}
+			m.state.ReqRenderFailure.Inc(1)
+			return bfe_module.BfeHandlerGoOn
+		}
+
+		m.state.ReqRenderSuccess.Inc(1)
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go]: fix in commit  dd7abf9

File: bfe_modules/mod_markdown/mod_markdown.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+				log.Logger.Debug("err in ModuleMarkdown.render.checkResponse(): %s", err)
+			}
+			return bfe_module.BfeHandlerGoOn
+		}
+		err = m.renderMarkDown(res)
+		if err != nil {
+			if openDebug {
+				log.Logger.Debug("err in ModuleMarkdown.renderMarkDown(): %s", err)
+			}
+			m.state.ReqRenderFailure.Inc(1)
```
[the remark this answers is not in the document]
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go]: fix in  dd7abf9

File: bfe_modules/mod_markdown/mod_markdown.go (hunk 1 above)
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go:146]: fix in commit  dd7abf9

File: bfe_modules/mod_markdown/mod_markdown.go (hunk 2 above)
Comment (YuqiXiao) [reply bfe_modules/mod_markdown/mod_markdown.go:153]: fix in commit  dd7abf9

**Pull request merged**